File: golang/internal/engine/diff.go
package engine

import (
	"context"
	"fmt"
	"strings"

	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"
)

// ExecuteChildWrite dispatches to the appropriate write mode handler.
func ExecuteChildWrite(ctx context.Context, q store.Querier, dialect store.Dialect, reg *metadata.Registry, parentID any, rw *RelationWrite) error {
	if rw.Relation.IsManyToMany() {
		return executeManyToManyWrite(ctx, q, dialect, reg, parentID, rw)
	}
	return executeOneToManyWrite(ctx, q, dialect, reg, parentID, rw)
}

func executeOneToManyWrite(ctx context.Context, q store.Querier, dialect store.Dialect, reg *metadata.Registry, parentID any, rw *RelationWrite) error {
	rel := rw.Relation
	targetEntity := reg.GetEntity(rel.Target)
	if targetEntity == nil {
		return fmt.Errorf("unknown target entity: %s", rel.Target)
	}

	switch rw.WriteMode {
	case "diff":
		return executeDiffWrite(ctx, q, dialect, targetEntity, rel, parentID, rw.Data)
	case "replace":
		return executeReplaceWrite(ctx, q, dialect, targetEntity, rel, parentID, rw.Data)
	case "append":
		return executeAppendWrite(ctx, q, dialect, targetEntity, rel, parentID, rw.Data)
	default:
		return executeDiffWrite(ctx, q, dialect, targetEntity, rel, parentID, rw.Data)
	}
}

func executeDiffWrite(ctx context.Context, q store.Querier, dialect store.Dialect, targetEntity *metadata.Entity, rel *metadata.Relation, parentID any, data []map[string]any) error {
	pkField := targetEntity.PrimaryKey.Field

	// Fetch current children
	existing, err := fetchCurrentChildren(ctx, q, dialect, targetEntity, rel, parentID)
	if err != nil {
		return err
	}
	existingByPK := indexByPK(existing, pkField)

	for _, row := range data {
		// Check for _delete flag
		if del, ok := row["_delete"]; ok && del == true {
			pk := row[pkField]
			if pk != nil {
				if err := softDeleteChild(ctx, q, dialect, targetEntity, pk); err != nil {
					return err
				}
			}
			continue
		}

		pk := row[pkField]
		if pk != nil {
			// Has PK — check if exists
			if _, exists := existingByPK[fmt.Sprintf("%v", pk)]; exists {
				// UPDATE
				if err := updateChild(ctx, q, dialect, targetEntity, pk, row); err != nil {
					return err
				}
			}
			// If PK provided but doesn't exist in current children, skip in diff mode
		} else {
			// No PK — INSERT
			row[rel.TargetKey] = parentID
			if err := insertChild(ctx, q, dialect, targetEntity, row); err != nil {
				return err
			}
		}
	}

	return nil
}

func executeReplaceWrite(ctx context.Context, q store.Querier, dialect store.Dialect, targetEntity *metadata.Entity, rel *metadata.Relation, parentID any, data []map[string]any) error {
	pkField := targetEntity.PrimaryKey.Field

	existing, err := fetchCurrentChildren(ctx, q, dialect, targetEntity, rel, parentID)
	if err != nil {
		return err
	}
	existingByPK := indexByPK(existing, pkField)
	seen := make(map[string]bool)

	for _, row := range data {
		pk := row[pkField]
		if pk != nil {
			pkStr := fmt.Sprintf("%v", pk)
			if _, exists := existingByPK[pkStr]; exists {
				seen[pkStr] = true
				if err := updateChild(ctx, q, dialect, targetEntity, pk, row); err != nil {
					return err
				}
			}
		} else {
			row[rel.TargetKey] = parentID
			if err := insertChild(ctx, q, dialect, targetEntity, row); err != nil {
				return err
			}
		}
	}

	// Soft-delete existing rows not in incoming
	for pkStr := range existingByPK {
		if !seen[pkStr] {
			pk := existingByPK[pkStr][pkField]
			if err := softDeleteChild(ctx, q, dialect, targetEntity, pk); err != nil {
				return err
			}
		}
	}

	return nil
}

func executeAppendWrite(ctx context.Context, q store.Querier, dialect store.Dialect, targetEntity *metadata.Entity, rel *metadata.Relation, parentID any, data []map[string]any) error {
	pkField := targetEntity.PrimaryKey.Field

	for _, row := range data {
		if row[pkField] != nil {
			continue // Skip rows with PK in append mode
		}
		row[rel.TargetKey] = parentID
		if err := insertChild(ctx, q, dialect, targetEntity, row); err != nil {
			return err
		}
	}
	return nil
}

// Many-to-many writes operate on the join table.
func executeManyToManyWrite(ctx context.Context, q store.Querier, dialect store.Dialect, reg *metadata.Registry, parentID any, rw *RelationWrite) error {
	rel := rw.Relation
	targetEntity := reg.GetEntity(rel.Target)
	if targetEntity == nil {
		return fmt.Errorf("unknown target entity: %s", rel.Target)
	}
	targetPKField := targetEntity.PrimaryKey.Field

	switch rw.WriteMode {
	case "replace":
		// Delete all current join rows, insert all incoming
		pb := dialect.NewParamBuilder()
		delSQL := fmt.Sprintf("DELETE FROM %s WHERE %s = %s", rel.JoinTable, rel.SourceJoinKey, pb.Add(parentID))
		if _, err := store.Exec(ctx, q, delSQL, pb.Params()...); err != nil {
			return fmt.Errorf("delete join rows: %w", err)
		}
		for _, row := range rw.Data {
			targetID := row[targetPKField]
			if targetID == nil {
				targetID = row["id"]
			}
			if targetID == nil {
				continue
			}
			if err := insertJoinRow(ctx, q, dialect, rel, parentID, targetID); err != nil {
				return err
			}
		}

	case "append":
		for _, row := range rw.Data {
			targetID := row[targetPKField]
			if targetID == nil {
				targetID = row["id"]
			}
			if targetID == nil {
				continue
			}
			// Insert only if not exists
			pb := dialect.NewParamBuilder()
			sql := fmt.Sprintf(
				"INSERT INTO %s (%s, %s) VALUES (%s, %s) ON CONFLICT DO NOTHING",
				rel.JoinTable, rel.SourceJoinKey, rel.TargetJoinKey, pb.Add(parentID), pb.Add(targetID))
			if _, err := store.Exec(ctx, q, sql, pb.Params()...); err != nil {
				return fmt.Errorf("insert join row: %w", err)
			}
		}

	default: // diff
		// Fetch current join rows
		pb := dialect.NewParamBuilder()
		currentSQL := fmt.Sprintf("SELECT %s FROM %s WHERE %s = %s",
			rel.TargetJoinKey, rel.JoinTable, rel.SourceJoinKey, pb.Add(parentID))
		currentRows, err := store.QueryRows(ctx, q, currentSQL, pb.Params()...)
		if err != nil {
			return fmt.Errorf("fetch current join rows: %w", err)
		}
		currentTargets := make(map[string]bool)
		for _, r := range currentRows {
			if v := r[rel.TargetJoinKey]; v != nil {
				currentTargets[fmt.Sprintf("%v", v)] = true
			}
		}

		for _, row := range rw.Data {
			targetID := row[targetPKField]
			if targetID == nil {
				targetID = row["id"]
			}
			if targetID == nil {
				continue
			}

			if del, ok := row["_delete"]; ok && del == true {
				dpb := dialect.NewParamBuilder()
				delSQL := fmt.Sprintf("DELETE FROM %s WHERE %s = %s AND %s = %s",
					rel.JoinTable, rel.SourceJoinKey, dpb.Add(parentID), rel.TargetJoinKey, dpb.Add(targetID))
				if _, err := store.Exec(ctx, q, delSQL, dpb.Params()...); err != nil {
					return fmt.Errorf("delete join row: %w", err)
				}
				continue
			}

			targetStr := fmt.Sprintf("%v", targetID)
			if !currentTargets[targetStr] {
				if err := insertJoinRow(ctx, q, dialect, rel, parentID, targetID); err != nil {
					return err
				}
			}
		}
	}

	return nil
}

// Helper functions

func fetchCurrentChildren(ctx context.Context, q store.Querier, dialect store.Dialect, entity *metadata.Entity, rel *metadata.Relation, parentID any) ([]map[string]any, error) {
	columns := strings.Join(entity.FieldNames(), ", ")
	sql := fmt.Sprintf("SELECT %s FROM %s WHERE %s = %s", columns, entity.Table, rel.TargetKey, dialect.Placeholder(1))
	if entity.SoftDelete {
		sql += " AND deleted_at IS NULL"
	}
	return store.QueryRows(ctx, q, sql, parentID)
}

func indexByPK(rows []map[string]any, pkField string) map[string]map[string]any {
	m := make(map[string]map[string]any, len(rows))
	for _, row := range rows {
		if pk := row[pkField]; pk != nil {
			m[fmt.Sprintf("%v", pk)] = row
		}
	}
	return m
}

func insertChild(ctx context.Context, q store.Querier, dialect store.Dialect, entity *metadata.Entity, fields map[string]any) error {
	sql, params := BuildInsertSQL(entity, fields, dialect)
	_, err := store.QueryRows(ctx, q, sql, params...)
	if err != nil {
		return fmt.Errorf("insert %s: %w", entity.Table, err)
	}
	return nil
}

func updateChild(ctx context.Context, q store.Querier, dialect store.Dialect, entity *metadata.Entity, id any, fields map[string]any) error {
	sql, params := BuildUpdateSQL(entity, id, fields, dialect)
	if sql == "" {
		return nil // nothing to update
	}
	if _, err := store.Exec(ctx, q, sql, params...); err != nil {
		return fmt.Errorf("update %s: %w", entity.Table, err)
	}
	return nil
}

func softDeleteChild(ctx context.Context, q store.Querier, dialect store.Dialect, entity *metadata.Entity, id any) error {
	if entity.SoftDelete {
		sql, params := BuildSoftDeleteSQL(entity, id, dialect)
		if _, err := store.Exec(ctx, q, sql, params...); err != nil {
			return fmt.Errorf("soft delete %s: %w", entity.Table, err)
		}
	} else {
		sql, params := BuildHardDeleteSQL(entity, id, dialect)
		if _, err := store.Exec(ctx, q, sql, params...); err != nil {
			return fmt.Errorf("hard delete %s: %w", entity.Table, err)
		}
	}
	return nil
}

func insertJoinRow(ctx context.Context, q store.Querier, dialect store.Dialect, rel *metadata.Relation, sourceID, targetID any) error {
	pb := dialect.NewParamBuilder()
	sql := fmt.Sprintf("INSERT INTO %s (%s, %s) VALUES (%s, %s)",
		rel.JoinTable, rel.SourceJoinKey, rel.TargetJoinKey, pb.Add(sourceID), pb.Add(targetID))
	if _, err := store.Exec(ctx, q, sql, pb.Params()...); err != nil {
		return fmt.Errorf("insert join row in %s: %w", rel.JoinTable, err)
	}
	return nil
}

File: backend/expressjs/src/engine/diff.ts
import type { Queryable } from "../store/postgres.js";
import { queryRows, exec } from "../store/postgres.js";
import type { Entity, Relation } from "../metadata/types.js";
import type { Registry } from "../metadata/registry.js";
import { fieldNames } from "../metadata/types.js";
import type { RelationWrite } from "./writer.js";
import {
  buildInsertSQL,
  buildUpdateSQL,
  buildSoftDeleteSQL,
  buildHardDeleteSQL,
} from "./writer.js";

export async function executeChildWrite(
  q: Queryable,
  registry: Registry,
  parentID: any,
  rw: RelationWrite,
): Promise<void> {
  if (rw.relation.type === "many_to_many") {
    await executeManyToManyWrite(q, registry, parentID, rw);
  } else {
    await executeOneToManyWrite(q, registry, parentID, rw);
  }
}

async function executeOneToManyWrite(
  q: Queryable,
  registry: Registry,
  parentID: any,
  rw: RelationWrite,
): Promise<void> {
  const rel = rw.relation;
  const targetEntity = registry.getEntity(rel.target);
  if (!targetEntity) {
    throw new Error(`Unknown target entity: ${rel.target}`);
  }

  switch (rw.writeMode) {
    case "diff":
      await executeDiffWrite(q, targetEntity, rel, parentID, rw.data);
      break;
    case "replace":
      await executeReplaceWrite(q, targetEntity, rel, parentID, rw.data);
      break;
    case "append":
      await executeAppendWrite(q, targetEntity, rel, parentID, rw.data);
      break;
    default:
      await executeDiffWrite(q, targetEntity, rel, parentID, rw.data);
  }
}

async function executeDiffWrite(
  q: Queryable,
  targetEntity: Entity,
  rel: Relation,
  parentID: any,
  data: Record<string, any>[],
): Promise<void> {
  const pkField = targetEntity.primary_key.field;
  const existing = await fetchCurrentChildren(q, targetEntity, rel, parentID);
  const existingByPK = indexByPK(existing, pkField);

  for (const row of data) {
    if (row._delete === true) {
      const pk = row[pkField];
      if (pk != null) {
        await softDeleteChild(q, targetEntity, pk);
      }
      continue;
    }

    const pk = row[pkField];
    if (pk != null) {
      if (existingByPK.has(String(pk))) {
        await updateChild(q, targetEntity, pk, row);
      }
      // PK provided but not in current children — skip in diff mode
    } else {
      row[rel.target_key!] = parentID;
      await insertChild(q, targetEntity, row);
    }
  }
}

async function executeReplaceWrite(
  q: Queryable,
  targetEntity: Entity,
  rel: Relation,
  parentID: any,
  data: Record<string, any>[],
): Promise<void> {
  const pkField = targetEntity.primary_key.field;
  const existing = await fetchCurrentChildren(q, targetEntity, rel, parentID);
  const existingByPK = indexByPK(existing, pkField);
  const seen = new Set<string>();

  for (const row of data) {
    const pk = row[pkField];
    if (pk != null) {
      const pkStr = String(pk);
      if (existingByPK.has(pkStr)) {
        seen.add(pkStr);
        await updateChild(q, targetEntity, pk, row);
      }
    } else {
      row[rel.target_key!] = parentID;
      await insertChild(q, targetEntity, row);
    }
  }

  // Soft-delete existing rows not in incoming
  for (const [pkStr, row] of existingByPK) {
    if (!seen.has(pkStr)) {
      await softDeleteChild(q, targetEntity, row[pkField]);
    }
  }
}

async function executeAppendWrite(
  q: Queryable,
  targetEntity: Entity,
  rel: Relation,
  parentID: any,
  data: Record<string, any>[],
): Promise<void> {
  const pkField = targetEntity.primary_key.field;

  for (const row of data) {
    if (row[pkField] != null) continue;
    row[rel.target_key!] = parentID;
    await insertChild(q, targetEntity, row);
  }
}

async function executeManyToManyWrite(
  q: Queryable,
  registry: Registry,
  parentID: any,
  rw: RelationWrite,
): Promise<void> {
  const rel = rw.relation;
  const targetEntity = registry.getEntity(rel.target);
  if (!targetEntity) {
    throw new Error(`Unknown target entity: ${rel.target}`);
  }
  const targetPKField = targetEntity.primary_key.field;

  switch (rw.writeMode) {
    case "replace": {
      const delSQL = `DELETE FROM ${rel.join_table} WHERE ${rel.source_join_key} = $1`;
      await exec(q, delSQL, [parentID]);
      for (const row of rw.data) {
        const targetID = row[targetPKField] ?? row["id"];
        if (targetID == null) continue;
        await insertJoinRow(q, rel, parentID, targetID);
      }
      break;
    }

    case "append": {
      for (const row of rw.data) {
        const targetID = row[targetPKField] ?? row["id"];
        if (targetID == null) continue;
        const sql = `INSERT INTO ${rel.join_table} (${rel.source_join_key}, ${rel.target_join_key}) VALUES ($1, $2) ON CONFLICT DO NOTHING`;
        await exec(q, sql, [parentID, targetID]);
      }
      break;
    }

    default: {
      // diff mode
      const currentSQL = `SELECT ${rel.target_join_key} FROM ${rel.join_table} WHERE ${rel.source_join_key} = $1`;
      const currentRows = await queryRows(q, currentSQL, [parentID]);
      const currentTargets = new Set<string>();
      for (const r of currentRows) {
        if (r[rel.target_join_key!] != null) {
          currentTargets.add(String(r[rel.target_join_key!]));
        }
      }

      for (const row of rw.data) {
        const targetID = row[targetPKField] ?? row["id"];
        if (targetID == null) continue;

        if (row._delete === true) {
          const delSQL = `DELETE FROM ${rel.join_table} WHERE ${rel.source_join_key} = $1 AND ${rel.target_join_key} = $2`;
          await exec(q, delSQL, [parentID, targetID]);
          continue;
        }

        if (!currentTargets.has(String(targetID))) {
          await insertJoinRow(q, rel, parentID, targetID);
        }
      }
    }
  }
}

// Helper functions

async function fetchCurrentChildren(
  q: Queryable,
  entity: Entity,
  rel: Relation,
  parentID: any,
): Promise<Record<string, any>[]> {
  const columns = fieldNames(entity).join(", ");
  let sql = `SELECT ${columns} FROM ${entity.table} WHERE ${rel.target_key} = $1`;
  if (entity.soft_delete) {
    sql += " AND deleted_at IS NULL";
  }
  return queryRows(q, sql, [parentID]);
}

function indexByPK(
  rows: Record<string, any>[],
  pkField: string,
): Map<string, Record<string, any>> {
  const m = new Map<string, Record<string, any>>();
  for (const row of rows) {
    if (row[pkField] != null) {
      m.set(String(row[pkField]), row);
    }
  }
  return m;
}

async function insertChild(
  q: Queryable,
  entity: Entity,
  fields: Record<string, any>,
): Promise<void> {
  const [sql, params] = buildInsertSQL(entity, fields);
  await queryRows(q, sql, params);
}

async function updateChild(
  q: Queryable,
  entity: Entity,
  id: any,
  fields: Record<string, any>,
): Promise<void> {
  const [sql, params] = buildUpdateSQL(entity, id, fields);
  if (!sql) return;
  await exec(q, sql, params!);
}

async function softDeleteChild(
  q: Queryable,
  entity: Entity,
  id: any,
): Promise<void> {
  if (entity.soft_delete) {
    const [sql, params] = buildSoftDeleteSQL(entity, id);
    await exec(q, sql, params);
  } else {
    const [sql, params] = buildHardDeleteSQL(entity, id);
    await exec(q, sql, params);
  }
}

async function insertJoinRow(
  q: Queryable,
  rel: Relation,
  sourceID: any,
  targetID: any,
): Promise<void> {
  const sql = `INSERT INTO ${rel.join_table} (${rel.source_join_key}, ${rel.target_join_key}) VALUES ($1, $2)`;
  await exec(q, sql, [sourceID, targetID]);
}
